File: java/DaryHeap.java
// complete D-ary tree where each node has at most D children
// generalization of a binary heap 

public class DaryHeap{
  private final int d;
  private final List<Integer> heap;
  private final boolean isMaxHeap;

  public DaryHeap(int d, boolean isMaxHeap){
    this.d = d;
    this.heap = new ArrayList<>();
    this.isMaxHeap = isMaxHeap;
  }
  private int getParentIndex(int childIndex) {
    return(childIndex - 1) / d;
  }
  private int getChildIndex(int parentIndex, int childNumber) {
    return d * parentIndex + childNumber + 1;
  }
  private void heapifyUp(int index) {
    int parentIndex = getParentIndex(index);
    if(parentIndex >= 0) {
      if(isMaxHeap) {
        if(heap.get(index) > heap.get(parentIndex)) {
          swap(index, parentIndex);
          heapifyUp(parentIndex);
        }
      } else {
        if (heap.get(index) < heap.get(parentIndex)) {
          swap(index, parentIndex);
          heapifyUp(parentIndex);
        }
      }
    }
  }
  private void heapifyDown(int index) {
    int largestOrSmallestIndex = index;
    for(int i = 0; i < d; i++) {
      int childIndex = getChildIndex(index, i);
      if(childIndex < heap.size()) {
        if(isMaxHeap) {
          if(heap.get(childIndex) > heap.get(largestOrSmallestIndex)) {
            largestOrSmallestIndex = childIndex;
          }
        } else {
          if(heap.get(childIndex) < heap.get(largestOrSmallestIndex)) {
            largestOrSmallestIndex = childIndex;
          }
        }
      }
    }
    if(largestOrSmallestIndex != index) {
      swap(index, largestOrSmallestIndex);
      heapifyDown(largestOrSmallestIndex);
    }
  }
  private void swap(int index1, int index2) {
    int temp = heap.get(index1);
    heap.set(index1, heap.get(index2));
    heap.set(index2, temp);
  }
  public void insert(int value) {
    heap.add(value);
    heapifyUp(heap.size() - 1);
  }
  public int extractMax() {
    if(heap.isEmpty()) {
      throw new IllegalStateException("heap is empty");
    }
    int max = heap.get(0);
    swap(0, heap.size() - 1);
    heap.remove(heap.size() - 1);
    heapifyDown(0);
    return max;
  }
  public int extractMin() {
    if(heap.isEmpty()) {
      throw new IllegalStateException("heap is empty");
    }
    int min = heap.get(0);
    swap(0, heap.size() - 1);
    heap.remove(heap.size() - 1);
    heapifyDown(0);
    return min;
  }
}

File: python/dary_heap.py
# use cases of d-ary heaps
# priority queues; for applications where number of children per node can vary
# sorting algorithms; can be used to implement heap sort 
# operations and their time complexity
# insert; O(log_d n)
# extract min/max; O(log_d n)



class DaryHeap:
  def __init__(self, d, heap_type='min'):
    self.heap = []
    self.d = d
    self.heap_type = heap_type

  def parent(self, i):
    return (i - 1) // self.d

  def child(self, i, j):
    return self.d * i + j + 1

  def heapify_up(self, i):
    while i > 0:
      p = self.parent(i)
      if self.heap[p] > self.heap[i] if self.heap_type == 'min' else self.heap[p] < self.heap[i]:
        self.heap[p], self.heap[i] = self.heap[i], self.heap[p]
        i = p
      else:
        break

  def heapify_down(self, i):
    while True:
      smallest = i
      for j in range(self, d):
        child = self.child(i, j)
        if child < len(self.heap) and self.heap[child] < self.heap[smallest] if self.heap_type == 'min' else self.heap[child] > self.heap[smallest]:
          smallest = child
      if smallest != i:
        self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
        i = smallest
      else:
        break

  def insert(self, value):
    self.heap.append(value)
    self.heapify_up(len(self.heap) - 1)

  def extract_min(self):
    if not self.heap:
      return None
    if len(self.heap) == 1:
      return self.heap.pop()
    root = self.heap[0]
    self.heap[0] = self.heap.pop()
    self.heapify_down(0)
    return root 

  def extract_max(self):
    if not self.heap:
      return None
    if len(self.heap) == 1:
      return self.heap.pop()
    root = self.heap[0]
    self.heap[0] = self.heap.pop()
    self.heapify_down(0)
    return root 
